scripts/python/generate_translations.py

```diff
@@ -0,0 +1,187 @@
+              "text": json.dumps(chunk, indent=2)
+            }
+          ]
+        }
+      ]
+    )
+    result = message.content[0].text
+
+    logging.debug(f'{locale} | Translation result: {result}')
+
+    json_match = regex.search(r'{(?:[^{}]|(?R))*}', result)
+    if not json_match:
+      raise ValueError('Failed to extract JSON from translation result.')
+
+    json_str = json_match.group()
+    # LLM may wrongly escape backslashes in the JSON string
+    json_str = regex.sub(r'\\{2}', r'\\', json_str)
+
+    translated_chunk = json_repair.loads(json_str)
+    translated_messages.update(translated_chunk) # type: ignore
+
+    logging.info(f'{locale} | Finished translation of chunk {idx + 1}/{len(chunks)}')
+
+    yield translated_messages
+
+def ensure_path_exists(path):
+  directory = os.path.dirname(path)
+  if not os.path.exists(directory):
+    os.makedirs(directory)
+
+  if not os.path.isfile(path):
+    with open(path, 'w'):
+        pass
+
+def update_translation(locale: str, default_translation_file: str, locale_translation_file: str, chunk_size):
+  logging.info(f'{locale} | Updating {locale_translation_file} with latest keys.')
+
+  with open(default_translation_file, 'r') as file:
+    default_translation = json.load(file, object_pairs_hook=collections.OrderedDict)
+
+  # Create locale translation file if it doesn't exist
+  ensure_path_exists(locale_translation_file)
+
+  with open(locale_translation_file, 'r') as file:
+    try:
+      current_translation = json.load(file, object_pairs_hook=collections.OrderedDict)
+    except json.JSONDecodeError:
+      current_translation = collections.OrderedDict()
+
+  missing_keys = find_missing_keys(default_translation, current_translation)
+
+  if len(missing_keys) == 0:
+    logging.info(f'{locale} | No missing keys found in {locale_translation_file}.')
+    return
+
+  for translated_messages in auto_translate(messages=missing_keys, locale=locale, chunk_size=chunk_size):
+    current_translation.update(translated_messages)
+    current_translation = collections.OrderedDict((key, current_translation[key]) for key in default_translation.keys() if key in current_translation)
+
+    with open(locale_translation_file, 'w') as file:
+      json.dump(current_translation, file, indent=2, ensure_ascii=False)
+      file.write('\n')
+
+  missing_keys = find_missing_keys(default_translation, current_translation)
+  if len(missing_keys) > 0:
+    logging.error(f'{locale} | Failed to translate the following keys: {missing_keys.keys()}.')
+
+  logging.info(f'{locale} | Updated {locale_translation_file} with latest keys.')
+
+
+def make_update_locale_fn(chunk_size):
+  def update_locale(locale):
+    try:
+      logging.info(f'{locale} | Starting translation for {locale} ({LOCALE_DICT[locale]})')
+
+      # HTML template translation
+      update_translation(
+        locale=locale,
+        default_translation_file=f'../../resources/authgear/templates/en/translation.json',
```

Comment:
This script should allow the user to choose the base language. For example, to generate the zh-CN translation, we want to use our zh-HK as the base language.